[Spaces] Space-Aware Saved Objects

legrego opened the pull request:
**This is an updated description that incorperates some of the discussions below**

This PR introduces changes which allow the Spaces plugin to make saved objects "space aware". Effectively, this means wrapping the Saved Objects Client to alter or filter requests/responses.

**Note** Advanced UI Settings (i.e., saved objects of type `config`) are not in scope for this PR, and will be addressed separately.


### Terminology:
- SOC: Saved Objects Client

## Saved Objects Client
### `get`
The response from the base SOC is checked to see if the object belongs to the current space. If not, a 404 is thrown to indicate the object does not exist.

### `bulk_get`
The response from the base SOC is checked to see if each object belongs to the current space. For each object that does not belong, its contents are replaced with a 404 response, which looks identical to the base SOC's 404 response for a missing object.

### `create`
The `spaceId` is appended to the create request, so the base SOC will write the new object into the correct space.

### `bulk_create`
The `spaceId` is appended to each space-aware object in the request, so the base SOC will write the new objects into the correct space.

### `update`
Before allowing an update to be processed by the base SOC, we check to ensure that it belongs to the current space. If not, a 404 is thrown. We also ensure that the `spaceId` is not changed as a result of an update.

### `delete`
Before allowing a delete to be processed by the base SOC, we check to ensure that it belongs to the current space. If not, a 404 is thrown.

### `find`
Searching is arguably the most complex case for this PR, and is responsible for a bulk of the LOC (other than tests). When performing a find, we augment the ES query to ensure that each object belongs to the current space.


## * Belonging to the current space
To figure out if an object belongs to the current space, the following check is performed:
### 1. Is the object's type space-aware?
Most saved object types are space-aware. There are a couple of exceptions as of this PR: space and config.
If the type is not space-aware, then ✅ this object belongs to the current space. This implies that objects that are not space aware belong to every space.

If the type is space-aware, then processing continues to step 2
### 2. Check the object's `spaceId`
Each saved object may have a `spaceId` assigned. This `spaceId` is compared against the `spaceId` that the user's request is executed within. If they match, then the object belongs to the current space.

**caveat** The Default Space is a special-case space that does not assign a `spaceId` to its underlying objects. This is done to maintain backwards compatibility, and makes bootstrapping Spaces much easier for upgrading installations. Given this, there is logic in place which accounts for this special-case. The most interesting example is when we build the query for the SOC's `find` operation. Rather than checking that the object has a particular `spaceId`, we have to check that the object does not have a `spaceId` assigned.

legrego commented on src/server/saved_objects/client/saved_objects_client.js:
```diff
@@ -122,7 +123,8 @@ export class SavedObjectsClient {
           updated_at: time,
-          [type]: attributes
+          [type]: attributes,
+          ...extraBodyProperties
```
Used to assign the `spaceId`

kobelb replied:
Should we put this first so it can't overwrite the other properties?

legrego commented on src/server/saved_objects/client/saved_objects_client.js:
```diff
@@ -169,7 +171,8 @@ export class SavedObjectsClient {
           updated_at: time,
-          [object.type]: object.attributes
+          [object.type]: object.attributes,
+          ...object.extraBodyProperties
```
Used to assign the `spaceId`

kobelb replied:
Let's move this one up "top" as well, like we did for `create`

legrego replied:
Thanks -- don't know how I missed that...fixed!

legrego commented on src/server/saved_objects/client/saved_objects_client.js:
```diff
@@ -300,6 +304,10 @@ export class SavedObjectsClient {
     };
 
+    if (esOptions.body.query && typeof queryDecorator === 'function') {
+      esOptions.body.query = queryDecorator(esOptions.body.query);
```
Augments the ES query to restrict results to the active Space

kobelb replied:
If we only support functions, let's change this to throw an Error if something else is provided instead of silently ignoring it.

legrego commented on src/server/saved_objects/client/saved_objects_client.js:
```diff
@@ -357,8 +365,15 @@ export class SavedObjectsClient {
+
+    let docsToReturn = docs;
+    if (typeof options.documentFilter === 'function') {
+      docsToReturn = docs.filter(options.documentFilter);
```
Filters ES results to only show documents for the active Space

legrego commented on src/server/saved_objects/client/saved_objects_client.js:
```diff
@@ -370,13 +385,19 @@ export class SavedObjectsClient {
         };
+
+        if (typeof options.resultDecorator === 'function') {
+          return options.resultDecorator(savedObject, doc);
```
Hacky...this might not be needed. Adds the `spaceId` to each returned object

legrego commented on src/server/saved_objects/client/saved_objects_client.js:
```diff
@@ -388,14 +409,18 @@ export class SavedObjectsClient {
     });
 
+    if (typeof options.responseInterceptor === 'function') {
+      options.responseInterceptor(response);
```
Filter the ES search result based on the active Space

kobelb commented on x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js:
```diff
@@ -0,0 +1,152 @@
+    if (this._isTypeSpaceAware(options.type)) {
+      const spaceId = await this._getSpaceId();
+
+      spaceOptions.queryDecorator = (query) => {
```
Instead of using the decorator here, could we pass in the additional filter and let the SavedObjectsClient know how to append it? It's less generic, but let's us be less concerned with the internal implementation of the query itself.

kobelb commented on x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js:
```diff
@@ -0,0 +1,152 @@
+    // ES 'mget' does not support queries, so we have to filter results after the fact.
+    const thisSpaceId = await this._getSpaceId();
+
+    return await this._client.bulkGet(objects, {
```
In a similar fashion to the above, how do you feel about passing in an `extraSourceProperties` array that we use to copy the additional properties from the document? Then we could just do the filter on the result without adding the extension point.

legrego replied:
I like it! Done!

kobelb commented on x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js:
```diff
@@ -0,0 +1,152 @@
+    let thisSpaceId;
+
+    if (this._isTypeSpaceAware(type)) {
+      thisSpaceId = await this._getSpaceId();
```
Same as above, if we added the `extraSourceProperties` then we could by default assign that property from the _source to the object and then just filter the response.

legrego replied:
I like it! Done!

kobelb commented on x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  async update(type, id, attributes, options = {}) {
+    return await this._client.update(type, id, attributes, options);
```
To be safe, let's specify the spaceId here always so that they don't accidentally write the document without a spaceId or to the wrong spaceId.

kobelb commented on src/server/saved_objects/client/saved_objects_client.js:
```diff
@@ -357,8 +366,17 @@ export class SavedObjectsClient {
+    const { docs } = response;
+
+    let docsToReturn = docs;
+    if (typeof options.documentFilter === 'function') {
```
Do we still need this?

legrego replied:
I should be able to pull this logic up into the `spaces_saved_object_client` now that we've introduced `extraSourceProperties`

kobelb replied:
awesome, thanks!

archanid commented:
> ## create
> Add `spaceId` to the object being created

This is only if the object is being created in a space. Otherwise, if we're in the global space (no-space-land), no such thing is added to the object, right?

-----

> ## get + bulkGet

~Changing internally to use `find` (search) is an interesting idea. I did an audit of where all `get` and `bulkGet` are currently used to see what would be affected, and it seems like we just need a collection of objects, similar to what `find` returns. Get/bulkGet is less complex than the typical `find` operation can be. So, looking at [esOptions](https://github.com/elastic/kibana/blob/master/src/server/saved_objects/client/saved_objects_client.js#L285-L301), everything that `get` requires [(index, type, id, ignore)](https://github.com/elastic/kibana/blob/master/src/server/saved_objects/client/saved_objects_client.js#L392-L397) exists except for `id`. `bulkGet` calls with [this](https://github.com/elastic/kibana/blob/master/src/server/saved_objects/client/saved_objects_client.js#L350-L358), which would require some transformation if it were to use `find`.~

Scrap all that. I think it's a bad idea to use `search`/`find` inside of our current `get` and `bulkGet`.
1. `Search` will be less efficient because it must look through _all_ shards, whereas `get` knows the shard based on `hash("DOCID") % NUM_OF_SHARDS`, which `search` doesn't do. So if we think we're optimizing something by using `search`, we're not.
2. Current users of `get` expect the latest version of the document, whereas users of `search` do not. Behind the scenes, the index is refreshed if needed before doing the `get`. This probably affects us less, but it could be a factor.

So that leaves doing a `get` before updates and deletes, which should be rather cheap.

--------

**High level question:** What about OLS metadata? This is just about Spaces. Have we thought about what needs to change for OLS to be supported in the current Saved Objects Client?

archanid commented:
The optimization discussion above only would apply to `mget` usage, right? `get` is always getting a single document?

legrego commented:
> This is only if the object is being created in a space. Otherwise, if we're in the global space (no-space-land), no such thing is added to the object, right?

We only add the spaceId when Spaces are enabled. The spaceId will be appended to every created object that is space aware...that is to say, anything that goes through the saved object client. The one exception is the Default Space. This is a special, system generated Space that cannot be deleted, and represents all objects that were created prior to Spaces existing. Users can also create new objects in this space, and when they do, I expect that we will not be storing the spaceId -- even though the default space technically has an id.

> [...] Scrap all that. After more research I think it's a bad idea to use search/find inside of our current get and bulkGet. [...]

This is a great insight, thank you for explaining the performance impacts here.
The theoritical options for get/mget that I was trying to describe are:
1. Have Elasticsearch filter the search based on spaceId
2. Have Elasticsearch return the results as it does today, and have the Saved Objects Client[/Service] filter the results based on spaceId before returning to the consumer.

Option 1 is not viable because the `get` and `mget` operations do not allow for that kind of filtering, and search/find could have a performance degredation.

Option 2 allows for fast retrieval from Elasticsearch, at the expense of additional work on the Kibana server. Kibana will be receiving and parsing objects that it has to discard. I don't expect that it'll be a huge dataset, but I suppose it's something to consider.

> High level question: What about OLS metadata? This is just about Spaces. Have we thought about what needs to change for OLS to be supported in the current Saved Objects Client?

We haven't implemented anything at this point, but I think we have a vague idea of what needs to happen. The Security Plugin will likely take similar steps to augment search/find queries, and to filter get/mget results.

legrego commented:
@kobelb @archanid 

I've pushed some substantial changes to this PR, which could use another look. I've updated the description above to reflect the functionality/implementation that I was striving for, so hopefully, that will help clarify things as you review this.

This should enable space-aware saved objects across all of Kibana, with the exception of Advanced UI Settings (i.e., saved objects of type `config`). Spaces are also not space-aware themselves since their configuration is a global Kibana concept.

This does not include "securable spaces" -- that will be addressed in a separate PR

kobelb commented on x-pack/test/spaces_api_integration/apis/saved_objects/lib/space_test_utils.js:
```diff
@@ -0,0 +1,23 @@
+  return urlContext ? `/s/${urlContext}` : ``;
+}
+
+// Spaces do not actually prefix the ID, but this simplifies testing positive and negative flows.
```
Sorry for being dense, would you mind elaborating on this?

legrego replied:
This comment will be outdated by https://github.com/legrego/kibana/pull/5, but here, I am prefixing each object id with its space id inside the es archive. This allows me to reuse the base id, so that we have less hard-coded ids within tests, while still letting me assert that I am returning the expected objects, and only the expected objects.

archanid commented:
@legrego I'm out for the next full week. Can this wait until then? I can commit to review.

legrego commented on x-pack/plugins/spaces/server/lib/create_spaces_service.js:
```diff
@@ -0,0 +1,33 @@
+
+  const contextCache = new WeakMap();
+
+  function getUrlContext(request, defaultContext = null) {
```
Yep, I think this is a holdover...I'll work on removing this

legrego commented on x-pack/plugins/spaces/server/lib/saved_objects_client/__snapshots__/spaces_saved_objects_client.test.js.snap:
```diff
@@ -0,0 +1,7 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`#delete does not allow an object to be deleted via a different space 1`] = `"not found"`;
```
Yep! The underlying `repository` throws 404/NotFound for each of these scenarios

legrego commented on x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js:
```diff
@@ -0,0 +1,173 @@
+  async find(options = {}) {
+    const spaceOptions = {};
+
+    let types = options.type || this._types;
```
Yes, I believe so: If no types are specified, then we need to filter against all known types, because we need to know which types are space-aware, and which aren't. Space-aware types need to be constrained to the current space, while non-space-aware types no not.

legrego replied:
Alternatively, we can say that the `find` operation will only return space-aware types, unless they explicitly request a non-space-aware type. I'd need to verify, but it might simplify the export process (I assuming we don't want to export all spaces when a user decides to export objects from a particular space)

legrego commented on x-pack/plugins/spaces/server/routes/api/v1/spaces.js:
```diff
@@ -204,6 +205,29 @@ export function initSpacesApi(server) {
   });
 
+  server.route({
+    method: 'PUT',
```
I don't have a great reason for this, so I'm happy to make this a POST instead of a PUT if that's more consistent with the rest of the system

legrego commented on x-pack/plugins/spaces/server/routes/api/v1/spaces.test.js:
```diff
@@ -148,4 +153,18 @@ describe('Spaces API', () => {
     });
   });
+
+  test('PUT space/{id}/select should respond with the new space location', async () => {
```
Great suggestion, will do!

legrego commented on src/server/saved_objects/service/lib/repository.test.js:
```diff
@@ -155,11 +155,12 @@ describe('SavedObjectsRepository', () => {
     });
 
```
good catch -- added!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6279/)

legrego commented:
@kobelb I think I addressed all of your feedback. I also incorperated legrego/kibana#6, which removes the overly complex queryParams merge that I originally had

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6374/)

legrego commented:
@archanid this should be ready for your review whenever you get the chance. Note that none of the spaces work will pass CI at this point, so don't let that deter you from reviewing this PR. Thanks again!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6979/)

legrego commented:
Following a zoom with @kobelb, we will be reworking this so that the Spaces Saved Objects Client no longer deals with types that are not space-aware. Types that are not space-ware, for at least the near term, must be managed by their own dedicated clients.

For example, Spaces themselves are not space-aware. In other words, you are able to view and manage all spaces regardless of which space you are currently in. To accommodate this, this PR will introduce a `SpacesClient` alongside the Spaces Saved Objects Client.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7466/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7537/)

legrego commented:
jenkins test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7553/)

legrego commented on src/server/saved_objects/service/lib/repository.js:
```diff
@@ -54,11 +54,13 @@ export class SavedObjectsRepository {
   async create(type, attributes = {}, options = {}) {
     const {
       id,
+      extraBodyProperties = {},
```
How about `extraDocumentProperties`?

kobelb replied:
Perfect!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7600/)

legrego commented:
@kobelb since this PR doesn't deal with securing access to Spaces, I'm inclined to leave this as-is (pending feedback), and introduce the Spaces Client within the PR that actually enforces security. This PR is already getting rather large, and I have a couple of other features waiting on this one.

legrego commented on src/server/saved_objects/service/saved_objects_client.js:
```diff
@@ -102,6 +102,7 @@ export class SavedObjectsClient {
    * @param {object} [options={}]
    * @property {string} [options.id] - force id on creation, not recommended
    * @property {boolean} [options.overwrite=false]
+   * @property {object} [options.extraBodyProperties={}] - extra properties to append to the document body, outside of the object's type property
```
oops! yup!

legrego commented on x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js:
```diff
@@ -0,0 +1,193 @@
+      ...options
+    };
+
+    if (shouldAssignSpaceId) {
```
great catch, thanks!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7918/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7979/)
